spec/index.md

```diff
@@ -4391,28 +4395,47 @@ Env = {
 }
 
 (
-  (A.mode = upgrade or A.mode = upgrade {skip_pre_upgrade = false})
+  (A.mode = upgrade U and U.skip_pre_upgrade ≠ true)
   Env1 = Env with {
     global_timer = S.global_timer[A.canister_id];
     canister_version = S.canister_version[A.canister_id];
   }
-  Old_module.pre_upgrade(Old_State, M.caller, Env1) = Return {stable_memory = Stable_memory; new_certified_data = New_certified_data; cycles_used = Cycles_used;}
+  Old_module.pre_upgrade(Old_State, M.caller, Env1) = Return {new_state = Intermediate_state; new_certified_data = New_certified_data; cycles_used = Cycles_used;}
 )
 or
 (
-  A.mode = upgrade {skip_pre_upgrade = true}
-  Stable_memory = Old_State.stable_mem
+  (A.mode = upgrade U and U.skip_pre_upgrade = true)
+  Intermediate_state = Old_state
   New_certified_data = NoCertifiedData
   Cycles_used = 0
 )
 
+(
+  (A.mode = upgrade U and U.wasm_memory_persistence ≠ keep)
+  Persisted_state = {store = initial_wasm_store; self_id = A.canister_id; stable_mem = Intermediate_state.stable_memory}
+)
+or
+(
+  (A.mode = upgrade U and U.wasm_memory_persistence = keep)
+  Persisted_state = initialize_store(Intermediate_state, A.wasm_module)
+)
+
+(A.mode = upgrade U and U.wasm_memory_persistence = keep)
+or
+(A.mode = upgrade U and U.wasm_memory_persistence = replace)
+or
+(S.canisters[A.canister_id].private_custom_sections["enhanced-orthogonal-persistence"] = null)
+
+not (A.mode = upgrade U and U.wasm_memory_persistence = keep and Private_custom_sections["enhanced-orthogonal-persistence"] = null)
```

Comment:
@luc-blaeser Should we add an analogous check that the old module must declare the custom section "enhanced-orthogonal-persistence" if we use the "keep" option?

Reply:
I just discussed this with @luc-blaeser and I think his implementation implements that check, so the spec is probably missing it. But best to get confirmation from @luc-blaeser himself...

Reply:
Yes, I agree, it would make sense to check that both at least the **new** version has the EOP custom section, when `Keep` is specified. Otherwise, a classical version would misinterpret the non-zero main memory and not see any serialized state in stable memory. This is currently implemented in the MR (https://gitlab.com/dfinity-lab/public/ic/-/merge_requests/17893)

We could also additionally require that the **old** version carries the EOP custom section, when `Keep` is specified. However, this is less critical, as the EOP runtime should detect this case and simply deserialize state from stable memory and reset the main memory state. This check is currently not implemented in the MR.

Reply:
>  This is currently implemented in the MR

and also in the spec at the line

```
not (A.mode = upgrade U and U.wasm_memory_persistence = keep and Private_custom_sections["enhanced-orthogonal-persistence"] = null)
```

> This check is currently not implemented in the MR and in my opinion, not necessarily needed.

This is what I was asking.

>  as the EOP runtime should detect this case and simply deserialize state from stable memory and reset the main memory state

Then it sounds safe, indeed.